<a href="https://colab.research.google.com/github/tabeeljohn2/flyrank-ml-internship/blob/main/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import os, subprocess

# Download YOUR repo into this Colab session (only if not already there)
if not os.path.exists("flyrank-ml-internship"):
    subprocess.run(
        ["git", "clone", "https://github.com/tabeeljohn2/flyrank-ml-internship.git"],
        check=True,
    )

# Move into the repo folder so relative paths like data/raw/... work
os.chdir("flyrank-ml-internship")
print("Now in:", os.getcwd())
print("Data file exists:", os.path.exists("data/raw/content_refresh_anonymized.csv"))

Now in: /content/flyrank-ml-internship
Data file exists: True


# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/tabeeljohn2/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

Provisional lane: Refresh / Content Opportunity Scoring.
I chose it because the starter notebooks already showed that a learned
ranking can beat a fixed hand rule for finding pages worth reviewing, and
the decision behind it (which page to review first) is clear and practical.
I may change this lane until the end of Week 4.

In [2]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

Search question: Which existing pages should a content team review first
because they may be losing search visibility?

Unit of analysis: one page (one row per page).

Output: a ranked review list, most at-risk pages first, with reason codes
explaining why each page ranks high.

Action: a content team reviews the top-ranked pages and decides whether to
refresh, expand, or monitor each one.

Cost of a wrong recommendation: if a healthy page is ranked high, the team
wastes limited review time. If a page that really needed attention is ranked
low, it keeps losing visibility unnoticed. A recommendation is a prompt for
human review, not a guarantee that editing the page will help.

Why data or ML can help: there are too many pages to review by hand, and in
Notebook 01 a simple hand-written rule found only about 24% correct in its
top 50, while a learned model scored higher. The value is in the decision it
improves (who gets reviewed first), not in the algorithm itself.

In [3]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [4]:
import pandas as pd

# The small starter dataset that ships with the repo
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

# Same filter the pipeline uses: pages with visibility and enough history
df = df[(df["impressions_90d"] > 0) & (df["content_age_days"] >= 90)]

# Label used in the starter pipeline (a proxy for decline)
df["is_declining_label"] = (df["trend_direction"] == "down")

# Number 1: how many pages
print(f"Pages in the starter dataset: {len(df):,}")

# Number 2: how common decline is
share_down = df["is_declining_label"].mean()
print(f"Share of pages trending down: {share_down:.1%}")

# Number 3: is there a stale but visible group worth reviewing?
stale_visible = ((df["days_since_last_update"] >= 180) & (df["impressions_90d"] >= 500)).sum()
print(f"Stale (180+ days) pages with 500+ impressions: {stale_visible:,}")


Pages in the starter dataset: 30,000
Share of pages trending down: 54.2%
Stale (180+ days) pages with 500+ impressions: 17


Observed: of 30,000 pages, 54.2% were trending down, so roughly half of the
inventory shows a decline signal, which is far too many pages to review by
hand. Only 17 pages were both stale (180+ days since update) and visible
(500+ impressions), which suggests a fixed rule like this catches a very
small group, while a learned ranking can prioritize across the whole set.
This is an observation on the starter dataset, not a claim about all clients.

Note: the "declining" label is a proxy based on the current window's
trend_direction, so 54.2% describes how the label is defined here, not
the true rate of decline.

## 4. Careful words: what I can and can't claim

What I can claim: in this anonymized starter dataset, I observed that a
learned model ranked declining pages better than a fixed rule, directionally.

What I cannot claim: that a refresh would cause recovery (that needs an
experiment), that this holds for other clients or the full warehouse, or
anything about how Google ranks pages. The label used in the starter data
is a proxy (trend_direction == "down"), not a future outcome.

In [5]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.